### PR TITLE
Overhaul large OD tool to resolve various problems

### PR DESCRIPTION
**Pull request by mmorang**
Full overhaul of tool. Fixes a couple of key things:
- Pipes logger messages to GP UI, including regular status updates (https://github.com/Esri/large-network-analysis-tools/issues/7)
- Better handles time and distance units and pre-selecting based on straight-line distance
- Puts OD settings in a config file for easy access
- Ensures proper post-processing to handle finding the closest K destinations so that when all chunks are stitched back together, the top K are kept.
- Add unit tests to ensure correctness
- Convert from tbx to pyt

### CODE REVIEW

File: unittests/unittests_odcm.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+import datetime
+import unittest
+from copy import deepcopy
+
+import arcpy
+
+CWD = os.path.dirname(os.path.abspath(__file__))
+sys.path.append(os.path.dirname(CWD))
+import odcm
+import portal_credentials  # Contains log-in for an ArcGIS Online account to use as a test portal
```
Comment (deelesh) [inline unittests/unittests_odcm.py:29]: @mmorang 
I prefer we include the portal_credentials.py file in the repo but without the credentials. The unit test should fail with invalid username and password error if the portal_credentials file does not contain valid credentials. 

I do not like the idea of having to do a lot more work in order to get the unit tests to run.

**Review by deelesh — COMMENTED** [review]
@mmorang 

I don't think the new code changes are following python style guide. I see a quite a few errors from pylint, pydocstyle and pycodestyle. When I wrote the initial version, I had ensured the code successfully lints when using the above three style checkers. 

I would prefer if the new code also follows the same convention. Can you install the three linters and fix all the errors / warnings?

File: LargeNetworkAnalysisTools.pyt [hunk 2]
```diff
@@ -0,0 +1,377 @@
+############################################################################
```
Comment (deelesh) [inline LargeNetworkAnalysisTools.pyt:1]: @mmorang 
I am not sure what is the motivation behind moving the toolbox from .tbx to .pyt. One downside I see is using the GP tool from the .pyt is missing the help for all the parameters. In a .pyt the help is stored in the .pyt.xml file which you are ignoring from the repo.

File: LargeNetworkAnalysisTools.pyt [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+
+        param_num_dests = arcpy.Parameter(
+            displayName="Number of Destinations to Find for Each Origin",
+            name="Num_Destinations",
+            datatype="GPLong",
+            parameterType="Optional",
+            direction="Input"
+        )
+
+        param_barriers = arcpy.Parameter(
```
Comment (deelesh) [inline LargeNetworkAnalysisTools.pyt:160]: @mmorang 
I had to look through the code to figure out that barriers parameter is supporting all three barrier types. Can we somehow make this more obvious? An easy way would have been if we have side panel help for this parameter.

**Review by deelesh — CHANGES_REQUESTED** [review]
@mmorang 
Looks good for most parts. I would love to run the tool with few inputs of my own before approving the PR. But for now I always get an error when running the tool. So I am waiting for a version that allows me to execute the tool.

File: LargeNetworkAnalysisTools.pyt (hunk 2 above)
Comment (mmorang) [reply LargeNetworkAnalysisTools.pyt:1]: Pyt makes it easier to have all the parameters defined in one place, see changes when syncing the repo, and have validation in the same place.

I completely forgot about the side panel help, though! Thanks for catching that. I have added it now.

File: LargeNetworkAnalysisTools.pyt (hunk 3 above)
Comment (mmorang) [reply LargeNetworkAnalysisTools.pyt:160]: I added side panel help.

**Comment by mmorang** [comment]
For some reason my linter had pylint turned off even though pycodestyle was working. Whatever. Anyway, I updated it to make linting happy, but if you're still seeing errors, please let me know what your VSCode settings are so I can get the same set-up.

This should be ready for a re-review now.

**Comment by mmorang** [comment]
FYI, further testing reveals that the tool will fail with a schema lock error when trying to calculate barrier location fields if the barriers are in the map. Normally that wouldn't happen, but it does in this case because Calculate Locations is run in a subprocess. I think the solution to this is to move all the preprocessing out of the subprocess entirely. The only part that should be in the subprocess is the part that launches the parallel processing and deals directly with the results of that.

That refactoring will take me a while, so you can hold off on your re-review until I do that.

**Comment by mmorang** [comment]
I'm going to close this. I refactored everything and ultimately got it all to work. I've preserved this branch for now just in case, but I'm going to close the PR. I created a new PR for the overhaul: https://github.com/Esri/large-network-analysis-tools/pull/9

**Pull request closed without merging**